### PR TITLE
Enhancement 12 add cook time to recipe table

### LINKED ISSUE #12: Need to add Cook time entry to recipe table
Add cook entry to recipe table, it will be used in cookbook / recipe carousels.

### PR DESCRIPTION
**Pull request by Limlight86**
closes #12 
Screen Shot 2019-05-26 at 7 02 41 PM
Screen Shot 2019-05-26 at 7 02 14 PM

### CODE REVIEW

File: app/helpers/recipes_helper.rb [hunk 1]
```diff
@@ -0,0 +1,2 @@
+module RecipesHelper
+end
```
Comment (andyweiss1982) [inline app/helpers/recipes_helper.rb]: Why are these here?  Probably artifacts of a something you generated but don’t actually need.

File: app/assets/stylesheets/recipes_index.css [hunk 2]
```diff
@@ -0,0 +1,20 @@
+table, th, td {
```
Comment (andyweiss1982) [inline app/assets/stylesheets/recipes_index.css]: Why not out these in recipes.scss?

File: app/models/category.rb [hunk 3]
```diff
@@ -0,0 +1,36 @@
+class Category < ApplicationRecord
+  has_many :recipe_categories
+  has_many :recipes, through: :recipe_categories
+
+  class << self #doing this makes it so that it only needs to refer the DB once and then saves it, improves performance.
+    def latin
```
Comment (andyweiss1982) [inline app/models/category.rb]: I’m not sure how I fee about this. What are you trying to achieve, exactly?   Do you really want to have to add a line of code every time you add a category to the database?   I’m not dead set against this, but need to better understand how it helps you meet your objectives.

File: app/views/active_storage/blobs/_blob.html.erb [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+
+  <figcaption class="attachment__caption">
+    <% if caption = blob.try(:caption) %>
+      <%= caption %>
+    <% else %>
+      <span class="attachment__name"><%= blob.filename %></span>
+      <span class="attachment__size"><%= number_to_human_size blob.byte_size %></span>
+    <% end %>
+  </figcaption>
+</figure>
```
Comment (andyweiss1982) [inline app/views/active_storage/blobs/_blob.html.erb]: What does this file do?

File: app/views/categories/index.html.erb [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+      <th>Delete?</th>
+    </tr>
+  </thead>
+  <tbody>
+    <% @categories.each do |category| %>
+      <tr>
+        <td><%= link_to category.name, category_path(category)%></td>
+        <td><%= link_to 'Edit', edit_category_path(category) %></td>
+        <td><%= link_to 'Destroy', category_path(category),
+              method: :delete,
```
Comment (andyweiss1982) [inline app/views/categories/index.html.erb]: Indent these to match where the link to starts

File: app/controllers/recipes_controller.rb [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    else
+      render 'edit'
+    end
+  end
+
+  def destroy
+    @recipe = Recipe.find(params[:id])
+    @recipe.destroy
+   
+    redirect_to '/recipes', notice: 'Recipe was successfully destroyed.'
```
Comment (andyweiss1982) [inline app/controllers/recipes_controller.rb]: recipes_url

File: app/views/categories/show.html.erb [hunk 7]
```diff
@@ -0,0 +1,20 @@
+<% if notice %>
+      <div>
+        <%= notice %>
+      </div>
```
Comment (andyweiss1982) [inline app/views/categories/show.html.erb]: Same comment as before re: indentation and placement of notice

File: app/views/categories/show.html.erb [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+<table>
+  <thead>
+    <tr>
+      <th>Name</th>
+    </tr>
+  </thead>
+  <tbody>
+      <tr>
+        <td><%= @category.name %></td>  
+      </tr>
```
Comment (andyweiss1982) [inline app/views/categories/show.html.erb]: Indentation mismatch

File: db/seeds.rb [hunk 9]
```diff
@@ -1,7 +1,58 @@
...
+             ]
+
+categories.each { |category| Category.create(category) }
+
+eggs = Recipe.create(
+               name: "Fried Eggs",
+               cooktime: 10,
+               servings: 4,
+               difficulty: 1,
+               calories: 94,
```
Comment (andyweiss1982) [inline db/seeds.rb]: This indentation is interesting.   What made to choose right here?

**Comment by Limlight86** [comment]
Thank you for reviewing the code so quickly and for all the comments. Im going address all of these first thing tomorrow, I will make sure to beautify the code with proper indentations and adress some fo the other things you pointed out.

Sorry, I was trying to make a bunch of separate commits to piece my parts of the code as I was working on it, kind of like what you have been showing us in class but when I made a new issue and switched branches it took all the changes with it. I will try to make sure this doesn't happen again, maybe you can point out some things on Tuesday that I can do to avoid this in the future.

File: app/controllers/categories_controller.rb [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+
+    if @category.save
+      redirect_to @category, notice: "Category was successfully created"
+    else
+      render 'new'
+    end
+  end
+
+  def edit
+    @category = Category.find(params[:id])
```
[the remark this answers is not in the document]
Comment (Limlight86) [reply app/controllers/categories_controller.rb]: Ok I will look into moving this, when I first wrote it I was just following the cookie cutter build, I was just making sure it all worked correctly.

File: app/controllers/categories_controller.rb [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+    else
+      render 'edit'
+    end
+  end
+
+  def destroy
+    @category = Category.find(params[:id])
+    @category.destroy
+   
+    redirect_to '/categories', notice: 'Category was successfully destroyed.'
```
[the remark this answers is not in the document]
Comment (Limlight86) [reply app/controllers/categories_controller.rb]: Im going to look into this again, I originally tried doing it like that but I must have been using the wrong syntax because it would give me errors.

File: app/controllers/recipes_controller.rb [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+    @recipe = Recipe.find(params[:id])
+  end
+
+  def new
+    @recipe = Recipe.new
+    @categories = Category.all
+  end
+
+  def create
+    @recipe = Recipe.new(recipe_params)
```
[the remark this answers is not in the document]
Comment (Limlight86) [reply app/controllers/recipes_controller.rb]: I believe it leaves them blank, I tried making a few with no categories and it was allowed. I actually was looking into putting in the validations since currently you can make an empty recipe. I suppose I wasn't thinking of anyone outside us using this aspect of our site but it would need to be way more robust in the "real world".

File: app/controllers/recipes_controller.rb [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+
+  def edit
+    @recipe = Recipe.find(params[:id])
+    @categories = Category.all
+  end
+
+  def update
+    @recipe = Recipe.find(params[:id])
+    @categories = Category.all
+    categories = params[:recipe][:categories].reject!(&:empty?)
```
[the remark this answers is not in the document]
Comment (Limlight86) [reply app/controllers/recipes_controller.rb]: I did this part with Tommy today, I was having issues editing the categories on a recipe, at first if you didnt mark the categories it would replace them with no categories. He helped me figure out a way to keep them from changing if no changes were made and to add new ones if you picked ones that werent there. He shortened up the variable since I was calling params[:recipe][:categories] often. I will look into this some more.

File: app/helpers/recipes_helper.rb (hunk 1 above)
Comment (Limlight86) [reply app/helpers/recipes_helper.rb]: alot of lines of code were added during my process of trying to make things work and installing gems. im not sure what this is either.

File: app/assets/stylesheets/recipes_index.css (hunk 2 above)
Comment (Limlight86) [reply app/assets/stylesheets/recipes_index.css]: I can do that, I did it like this because of not knowing exactly how to work with the scss files, some of the old coding habits from doing react I suppose.

File: app/models/category.rb (hunk 3 above)
Comment (Limlight86) [reply app/models/category.rb]: This is is something that is left over from when the TA Nick was helping me on Saturday. He seemed to know alot about ruby on rails and backend stuff so he was showing me how this would make the app work smoother and more efficient. Now that we have the ability to add categories I can see how this would be annoying. I will look into this to see what can be done or maybe tone down this complexity.

File: app/views/active_storage/blobs/_blob.html.erb (hunk 4 above)
Comment (Limlight86) [reply app/views/active_storage/blobs/_blob.html.erb]: I believe this file was made in consequence of me uploading a picture into the action_text field and creating an entry in the recipes model. I only ever used that same picture but Im not sure if that would have changed anything. I did some reading but wasnt exactly keen on my understanding of it. Am I on the right track?

File: app/controllers/recipes_controller.rb (hunk 6 above)
Comment (Limlight86) [reply app/controllers/recipes_controller.rb]: Will do!

File: db/seeds.rb (hunk 9 above)
Comment (Limlight86) [reply db/seeds.rb]: I was going along with the left align, I do remember seeing you make all the spacing equal when doing something like this. I will make sure to practice that from now on.

**Comment by Limlight86** [comment]
I keep thinking of the database like working with the api, I included the image_url and video_url just incase we needed them to populate fields that need them. I wasnt too sure how well working with action_text and videos would be so I included the url so it can be referenced later if needed. My code has been guided by the parameters of the project, Im not thinking on a bigger scope of how it would behave when being handled by other people, I guess Im just trying to get things to work so we can have a good presentation, I will try to keep these things in mind, I always want to learn better practices and for those habits early.

File: app/views/categories/index.html.erb [hunk 14]
```diff
@@ -0,0 +1,28 @@
+<% if notice %>
+      <div>
+        <%= notice %>
+      </div>
+    <% end %>
```
[the remark this answers is not in the document]
Comment (Limlight86) [reply app/views/categories/index.html.erb]: Fixed and consolidated the code into application.

File: app/assets/stylesheets/application.css [hunk 15]
```diff
@@ -1,3 +1,4 @@
+
```
[the remark this answers is not in the document]
Comment (Limlight86) [reply app/assets/stylesheets/application.css]: Removed

File: app/assets/stylesheets/recipes_index.css [hunk 16]
```diff
@@ -0,0 +1,20 @@
...
+  border: 1px solid black;
+}
+
+table {
+  width: 100%;
+}
+
+th {
+  height: 50px;
+}
```
[the remark this answers is not in the document]
Comment (Limlight86) [reply app/assets/stylesheets/recipes_index.css]: gave all admin tables a class of admin_table.

File: app/controllers/categories_controller.rb (hunk 10 above)
[the remark this answers is not in the document]
Comment (Limlight86) [reply app/controllers/categories_controller.rb]: code has been consolidated

File: app/controllers/categories_controller.rb [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+  end
+
+  private
+  def category_params
```
[the remark this answers is not in the document]
Comment (Limlight86) [reply app/controllers/categories_controller.rb]: done.

File: app/assets/stylesheets/recipes_index.css (hunk 2 above)
Comment (Limlight86) [reply app/assets/stylesheets/recipes_index.css]: moved and deleted old css file. I noticed I have a categories.scss as well but all the recipe css is being applied to those tables as well. are the css files shared similar to react? could it be that the tables have the same class so thats why they are sharing the styles?

File: app/models/category.rb (hunk 3 above)
Comment (Limlight86) [reply app/models/category.rb]: I removed all these lines of code, I believe they weren't doing anything anyways but I am interested on learning more to reasoning Nick mentioned this, I know it makes the app perform better, especially when there could be hundreds of database calls.

File: app/views/categories/index.html.erb (hunk 5 above)
Comment (Limlight86) [reply app/views/categories/index.html.erb]: done

File: app/views/categories/show.html.erb (hunk 7 above)
Comment (Limlight86) [reply app/views/categories/show.html.erb]: done

File: app/views/categories/show.html.erb (hunk 8 above)
Comment (Limlight86) [reply app/views/categories/show.html.erb]: done

File: db/seeds.rb (hunk 9 above)
Comment (Limlight86) [reply db/seeds.rb]: redid the indentation, tried to clean it up and make the spacing better.

File: app/controllers/categories_controller.rb (hunk 11 above)
[the remark this answers is not in the document]
Comment (Limlight86) [reply app/controllers/categories_controller.rb]: updated.

File: app/controllers/recipes_controller.rb (hunk 6 above)
Comment (Limlight86) [reply app/controllers/recipes_controller.rb]: updated.

File: app/controllers/recipes_controller.rb (hunk 12 above)
[the remark this answers is not in the document]
Comment (Limlight86) [reply app/controllers/recipes_controller.rb]: Can we review this concern a little further when we meet tomorrow, so I can hear more on your train of thought, thanks!

File: app/controllers/recipes_controller.rb (hunk 13 above)
[the remark this answers is not in the document]
Comment (Limlight86) [reply app/controllers/recipes_controller.rb]: Can we review this concern a little further when we meet tomorrow, so I can hear more on your train of thought. This part was one of the harder parts for me to figure out and Im apprehensive to change anything here that might break the code. Thanks!

File: app/models/category.rb (hunk 3 above)
Comment (Limlight86) [reply app/models/category.rb]: Ok I was wrong, removing this broke the app when I tried to seed it again doing db:seed. Im trying to reword the other parts of my code that are throwing the errors but so far I havent been able to fix it...I might have to add it back and then look at it with you tomorrow.

**Comment by andyweiss1982** [comment]
No worries, this is great code! There’s a lot to be proud of here. Let touch base tomorrow and get it merged!

File: app/models/category.rb (hunk 3 above)
Comment (Limlight86) [reply app/models/category.rb]: Alright I figured it out, I had to redo the syntax when assigning the category ids to the recipes, I was doing it right but I had an underscore in my name string when it had to be a space... 🤦‍♂ Will be committing the changes now

**Comment by Limlight86** [comment]
Thank you sir! Enjoy the rest of your memorial day, see you tomorrow. Theres still cold brew, we didnt drink it all, so we can start the day off right!
